[cpp-ue4] bug-fix issue 10205: response json parsing logging a error to unreal console when there is no actual parsing problem

leith-bartrich opened the pull request:
This is my proposed fix to #10205

In summary, rather than walk all vars, it walks required vars and non-required (optional) vars.  And within the required vars, it differentiates between writeOnly and non-writeOnly vars.  It allows optional or writeOnly vars to fail to parse (because they're absent) and not trigger a big red error log indication of a parsing failure.  To be clear:  The generated code should not yell in red about an error, when a server is actually following the schema.  The balance here, is that it tries to parse the vars anyway, but doesn't yell unless the schema truly indicates their absence is critical.

There are many deficiencies here.  e.g.  A TryGetJsonValue call doesn't differentiate between a parse failure versus the absence of a value, in its return.  Which forces a lowest common denominator solution here.  e.g. It's not necessarily providing much runtime feedback where it could.  However, I think both of those deficiencies are actually a design change, rather than a bug fix.  They could be handled in a different issue and and discussed there.

Here, I'm only seeking to keep the generated code from logging errors for missing vars, that probably should be missing.

I moved the comments to keep the code looking nicer, since my original bug submission.

This change hasn't been tested against a large array of schemas.  But it has been tested against the simple schema provided in the bug, against a dirt simple Django backend.

The code generated from the samples seems sane to me. But I'm not as familiar with that schema.  Of note:  the pet.yml model is really the only one that marks certain fields required for a response.  However, this is why I think it should instead be logging more useful information to Verbose and VeryVerbose on a per field basis, rather than declare a ParsingFailure on the whole response object.

tagging: @Kahncode

<!-- Please check the completed items below -->
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

leith-bartrich commented:
A rebase to development HEAD and subsequent rebuild and generation shows the Docs and Samples CI issue might be with development HEAD rather than this request.

evidence after doing a rebase and generation locally:

```
git diff
diff --git a/samples/server/petstore/haskell-yesod/.openapi-generator/FILES b/samples/server/petstore/haskell-yesod/.openapi-generator/FILES
index 1f055028bf6..08edcd6331e 100644
--- a/samples/server/petstore/haskell-yesod/.openapi-generator/FILES
+++ b/samples/server/petstore/haskell-yesod/.openapi-generator/FILES
@@ -1,6 +1,5 @@
 .dir-locals.el
 .gitignore
-.openapi-generator-ignore
 README.md
 app/DevelMain.hs
 app/devel.hs
@@ -23,8 +22,5 @@ src/Settings.hs
 src/Settings/StaticFiles.hs
 stack.yaml
 static/.gitkeep
-test/Handler/PetSpec.hs
-test/Handler/StoreSpec.hs
-test/Handler/UserSpec.hs
 test/Spec.hs
 test/TestImport.hs
```

likely culprit: #10193

wing328 commented:
@leith-bartrich the build failure issue due to outdated doc/samples has been fixed in the master. Please rebase or merge latest master into your branch to resolve the issue.

wing328 commented:
FYI. https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword contains keywords to close an issue automatically when the PR is merged.

leith-bartrich commented:
Rebased, rebuilt, regenerated and pushed.  The docs and samples passed fine this time.  I'm unclear as to if the ci/circleci failure is meaningful.  It seems to be the parallel test that failed?

trying out your suggested auto issue link:
Fixes #10205 

Also, @wing328 , Do you want me to squash commits in this kind of pull requests?  Or leave them broken out?  I already squashed some silly ones that were just me fixing white-space mistakes.  But what about intermediate changes like moving comments around and such?

Kahncode commented on modules/openapi-generator/src/main/resources/cpp-ue4/model-source.mustache:
```diff
@@ -175,9 +175,20 @@ bool {{classname}}::FromJson(const TSharedPtr<FJsonValue>& JsonValue)
+	{{/requiredVars}}
+	{{#optionalVars}}
+	//Optional.  Its absence would be okay.
+	TryGetJsonValue(*Object, TEXT("{{baseName}}"), {{name}});
```
Normally this is handed by the TOptional used with optional vars, so this line should not have to be changed. Can you confirm?

leith-bartrich replied:
Noted.  Will change and test.

Kahncode commented on modules/openapi-generator/src/main/resources/cpp-ue4/model-source.mustache:
```diff
@@ -175,9 +175,20 @@ bool {{classname}}::FromJson(const TSharedPtr<FJsonValue>& JsonValue)
+	TryGetJsonValue(*Object, TEXT("{{baseName}}"), {{name}});
+    {{/isWriteOnly}}
+    {{^isWriteOnly}}
+	//Required and not write-only.  Its absence is likely a problem.
```
I would rather not have the comments in the generated code to be honest.

leith-bartrich replied:
Noted.  Will change.

Kahncode commented on modules/openapi-generator/src/main/resources/cpp-ue4/model-source.mustache:
```diff
@@ -175,9 +175,20 @@ bool {{classname}}::FromJson(const TSharedPtr<FJsonValue>& JsonValue)
 	bool ParseSuccess = true;
 
-    {{#vars}}
+	{{#requiredVars}}
```
WriteJson uses #vars and #required to perform something similar here, if possible i'd like to keep things consistent, so if it's possible to iterate on all the vars with #vars rather than #requiredVars and #optionalVars, this would put me more at ease, and be more consistent with the place those variables are declared in the first place. Alternatively, change where those variables are declared in the header, but I'd suggest the first approach.

leith-bartrich replied:
Implemented request change and tested.  Works fine.  Will push shortly.

leith-bartrich commented on modules/openapi-generator/src/main/resources/cpp-ue4/model-source.mustache:
```diff
@@ -175,9 +175,20 @@ bool {{classname}}::FromJson(const TSharedPtr<FJsonValue>& JsonValue)
+    {{/isWriteOnly}}
+    {{^isWriteOnly}}
+	//Required and not write-only.  Its absence is likely a problem.
 	ParseSuccess &= TryGetJsonValue(*Object, TEXT("{{baseName}}"), {{name}});
```
Just to clarify:  the red highlight on the & symbol  in the full &= operator is a syntax highlighting thing with GitHub.  If you look the generated sample for the petshop that is part of this commit, you'll see it does generate the full &= operator.

If that's not what you're commenting on, please clarify, I'm confused.

Kahncode replied:
Indeed, ignore my previous comment, i'm not used to github reviews.

leith-bartrich commented:
Requested changes made, generated samples and docs, and pushed.

One outstanding request remains based on what may be a misunderstanding on syntax highlighting.

Testing locally on my codebase fixes the errant behavior in a satisfactory manner.

Of note: the petstore sample doesn't seem to show any change here because it doesn't seem to use any writeOnly fields in its components.  Which is likely why this case never came up.

Kahncode commented on modules/openapi-generator/src/main/resources/cpp-ue4/model-source.mustache:
```diff
@@ -175,9 +175,19 @@ bool {{classname}}::FromJson(const TSharedPtr<FJsonValue>& JsonValue)
+	{{#isWriteOnly}}
+	TryGetJsonValue(*Object, TEXT("{{baseName}}"), {{name}});
+	{{/isWriteOnly}}
+	{{^isWriteOnly}}
```
```suggestion
	{{^isWriteOnly}} ParseSuccess &= {{/isWriteOnly}} TryGetJsonValue(*Object, TEXT("{{baseName}}"), {{name}});
```

Slightly more concise version, otherwise the change looks great.

leith-bartrich replied:
change made locally, tested and pushed.

leith-bartrich commented:
Made requested change for conciseness.  Tested against local codebase.  Regenerated samples and docs.  Pushed.

Kahncode commented:
Looks good for me. Thanks for the contribution!
